Get placeholder's information from the slide and master layouts

rabi92 opened the pull request:
### This PR includes
- Get slide and master layout information, create a mapping table for the nodes in the layouts, based on id, `idx` and `type`(Which is found on the placeholder type element, contained within `p:ph` tag). 
- Find the node of the powerpoint element in the layout tables. Referred as `spNode`
- Rewrite getting the positions logic to see if we don't find the element position, then we check from the slide layouts, if not then get from master layouts.
- Earlier the elements were removed from the resulting json if the positions were not found. Keep the placeholder elements even if the position is not found in layouts.
- exclude header, footers and slide-number type of placeholders, in-fact include only the placeholders that we support, which are `title, subtitles, body, pictures, tables and lists`
- Parse the placeholders lists.
- Remove extra paragraph contents added
- Remove empty list content.

#### Next Todos:
- Lists identification from the placeholders does not feel concrete, but the implementation right now cover all the cases that we have found. Throughout time we can see if there is any problems regarding this, which will result in making this identification better or to have more confidence in it. 
- Any other information that we might have and need from the placeholders, other than the position and list identification, may be required in the future, we will see it upon the use case or when the requirement rises.

_General knowledge on the placeholders and its types: http://officeopenxml.com/prSlide.php_

rabi92 committed:
fix: improve function to handle edge cases

getAttributeByPath to handle if there are objects within objects

rabi92 committed:
feat: parse slide layouts and create ids table for layout content

rabi92 committed:
feat: parse slide master layout and its id table

rabi92 committed:
feat: get relevant node from the placeholders mapping tables

rabi92 committed:
feat: get placeholders positions from layout if not found on element

rabi92 committed:
feat: remove non supported placeholders elements

rabi92 committed:
fix: return default text properties, if properties not found in ppt

rabi92 committed:
Merge branch 'dev' into placeholder-positions-layouts-parser

rabi92 committed:
feat: do not add empty paragraph contents and list items contents

rabi92 committed:
feat: support for the placeholder lists

rabi92 committed:
refactor: generic function for getting the position

rabi92 committed:
fix: normal placeholder text identifying as lists

rabi92 committed:
chore: update version

rabi92 committed:
chore: update version

shak commented on ts/parsers/elementparser.ts:
```diff
@@ -7,49 +7,134 @@ import * as isEmpty from "lodash.isempty";
+                this.element,
+                '["p:nvSpPr"][0]["p:nvPr"][0]["p:ph"][0]["$"]["type"]'
+            );
+            if (!SUPPORTED_PLACEHOLDERS.includes(type)) {
```
```suggestion
            if (SUPPORTED_PLACEHOLDERS.includes(type) === false) {
```

shak replied:
returns a definite boolean

rabi92 replied:
done

shak commented on ts/parsers/elementparser.ts:
```diff
@@ -7,49 +7,134 @@ import * as isEmpty from "lodash.isempty";
+                this.element,
+                '["p:nvSpPr"][0]["p:nvPr"][0]["p:ph"][0]["$"]["idx"]'
+            );
+            let slideLayoutSpNode = undefined;
+            slideLayoutSpNode = slideLayoutTables["idxTable"][placeholderIdx];
```
`slideLayoutSpNode` isn't reassigned so these lines can be combined and `slideLayoutSpNode` changed to `const`

rabi92 replied:
done

shak commented on ts/parsers/elementparser.ts:
```diff
@@ -7,49 +7,134 @@ import * as isEmpty from "lodash.isempty";
+        if (type !== undefined) {
+            slideLayoutSpNode = slideLayoutTables["typeTable"][type];
+            slideMasterSpNode = slideMasterTables["typeTable"][type];
+            return { slideLayoutSpNode, slideMasterSpNode };
```
```suggestion

            return { slideLayoutSpNode, slideMasterSpNode };
```

rabi92 replied:
done

rabi92 replied:
but for some reason its not showing outdated, but i doubled check, i added the new line

shak commented on ts/parsers/elementparser.ts:
```diff
@@ -7,49 +7,134 @@ import * as isEmpty from "lodash.isempty";
+        if (idx !== undefined) {
+            slideLayoutSpNode = slideLayoutTables["idxTable"][idx];
+            slideMasterSpNode = slideMasterSpNode["idxTable"][idx];
+            return { slideLayoutSpNode, slideMasterSpNode };
```
```suggestion

            return { slideLayoutSpNode, slideMasterSpNode };
```

rabi92 replied:
done

shak commented on ts/parsers/elementparser.ts:
```diff
@@ -7,49 +7,134 @@ import * as isEmpty from "lodash.isempty";
+            slideMasterSpNode = slideMasterSpNode["idxTable"][idx];
+            return { slideLayoutSpNode, slideMasterSpNode };
+        }
+        return { slideLayoutSpNode, slideMasterSpNode };
```
```suggestion

        return { slideLayoutSpNode, slideMasterSpNode };
```

rabi92 replied:
done

shak commented on ts/parsers/elementparser.ts:
```diff
@@ -7,49 +7,134 @@ import * as isEmpty from "lodash.isempty";
+    }
+
+    public getLayoutSpNodes(slideLayoutTables, slideMasterTables) {
+        const idx =
+            getAttributeByPath(this.element, ["p:nvSpPr", "p:nvPr", "p:ph"]) === undefined
+                ? undefined
+                : getAttributeByPath(this.element, ["p:nvSpPr", "p:nvPr", "p:ph", "$", "idx"]);
+        const type =
+            getAttributeByPath(this.element, ["p:nvSpPr", "p:nvPr", "p:ph"]) === undefined
+                ? undefined
+                : getAttributeByPath(this.element, ["p:nvSpPr", "p:nvPr", "p:ph", "$", "type"]);
```
```suggestion
        let idx = undefined;
        let type = undefined;
        
        if (getAttributeByPath(this.element, ["p:nvSpPr", "p:nvPr", "p:ph"]) !== undefined) {
            idx = getAttributeByPath(this.element, ["p:nvSpPr", "p:nvPr", "p:ph", "$", "idx"]);
            type = getAttributeByPath(this.element, ["p:nvSpPr", "p:nvPr", "p:ph", "$", "type"]);
        }
```

ternary's can be gotten rid off (saity check by change before comitting directly, I am reviewing this with a lot distractions)

rabi92 replied:
check was not efficient, removed ternary, getting values directly now, if we don't find it we will have undefined, and checks for undefined values is already there in the code below

shak commented on ts/parsers/elementparser.ts:
```diff
@@ -7,49 +7,134 @@ import * as isEmpty from "lodash.isempty";
+        );
+
+        const xfrmNodePath = '["p:spPr"][0]["a:xfrm"][0]';
+        const slideXfrmNode = getValueAtPath(this.element, xfrmNodePath);
+        const slideLayoutXfrmNode = getValueAtPath(slideLayoutSpNode, xfrmNodePath);
+        const slideMasterXfrmNode = getValueAtPath(slideMasterSpNode, xfrmNodePath);
+
+        if (slideXfrmNode) {
+            return this.getXfrmNodePosition(slideXfrmNode);
+        }
+
+        if (slideLayoutXfrmNode) {
+            return this.getXfrmNodePosition(slideLayoutXfrmNode);
+        }
+
+        if (slideMasterXfrmNode) {
+            return this.getXfrmNodePosition(slideMasterXfrmNode);
+        }
```
I guess you don't have resolve all three in the begining, you can delay resolving layout x frmNode until it is required, and then delay from master even more etc. .. slightly more efficient.

```
const slideXfrmNode = getValueAtPath(this.element, xfrmNodePath);
if (slideXfrmNode) {
       return this.getXfrmNodePosition(slideXfrmNode);
}

const slideLayoutXfrmNode = getValueAtPath(slideLayoutSpNode, xfrmNodePath);
if (slideLayoutXfrmNode) {
        return this.getXfrmNodePosition(slideLayoutXfrmNode);
}
...
```

rabi92 replied:
Good point, changed it

shak commented on ts/parsers/paragraphparser.ts:
```diff
@@ -224,18 +232,27 @@ export default class ParagraphParser {
+            size: 1200,
+            fontAttributes: [],
+            font: "Helvetica",
+            fillColor: "000000"
```
should it not be `#000000`

rabi92 replied:
mm yea, but we are not using colors, and its been like that in the initial code, if i change i have to change it in many places. 
so skipping this change for now, if that's okay with you ?

shak replied:
Yeah fair point, OK

shak commented on ts/parsers/slideParser.ts:
```diff
@@ -1,14 +1,141 @@
+        // @resName: ppt/slides/_rels/slide1.xml.rels
+        let relationshipArray = slideRelations["Relationships"]["Relationship"];
+        let layoutFilename = "";
+        if (relationshipArray.constructor === Array) {
```
`Array.isArray(relationshipArray)` doesn't work? There's also lodash array check if you have access to it already

rabi92 replied:
changed it to isArray, i don't have the lodash one installed. so using isArray now

shak commented on ts/parsers/slideParser.ts:
```diff
@@ -1,14 +1,141 @@
+        if (relationshipArray.constructor === Array) {
+            for (const relationship of relationshipArray) {
+                if (
+                    relationship["$"]["Type"] ===
+                    "http://schemas.openxmlformats.org/officeDocument/2006/relationships/slideLayout"
```
```suggestion
                    relationship["$"]["Type"] ===
                        "http://schemas.openxmlformats.org/officeDocument/2006/relationships/slideLayout"
```

shak replied:
This schema can also be a const defined in consts file if there is one, or as a class const.

rabi92 replied:
made a constants file and moved it there.

shak commented on ts/parsers/slideParser.ts:
```diff
@@ -1,14 +1,141 @@
+            for (const relationship of relationshipArray) {
+                if (
+                    relationship["$"]["Type"] ===
+                    "http://schemas.openxmlformats.org/officeDocument/2006/relationships/slideMaster"
```
this can also be a const

rabi92 replied:
done

shak commented on ts/parsers/slideParser.ts:
```diff
@@ -1,14 +1,141 @@
+        const slideLayoutResContent = await ZipHandler.parseSlideAttributes(slideLayoutResFilename);
+        relationshipArray = slideLayoutResContent["Relationships"]["Relationship"];
+        let masterFilename = "";
+        if (relationshipArray.constructor === Array) {
```
check this as well if a cleaner array check can be done

rabi92 replied:
using `isArray` now

shak commented on ts/parsers/slideParser.ts:
```diff
@@ -1,14 +1,141 @@
+                        const nvSpPrNode = node["p:nvSpPr"];
+                        const id = getAttributeByPath(nvSpPrNode[0], ["p:cNvPr", "$", "id"]);
+                        const idx = getAttributeByPath(nvSpPrNode[0], [
+                            "p:nvPr",
+                            "p:ph",
+                            "$",
+                            "idx"
```
there appears to be some inconsistency, sometimes these elements are in line and sometimes line break after each .. just check and make them consistent please

rabi92 replied:
some formatting got off, most likely when i worked with on the mac, and settings did not sync. i have reformatted all the files and have set the print width to 120. 
I have changed it in the vscode settings and have pushed it. so it doesn't happen again

shak commented on ts/helpers/attributesHandler.ts:
```diff
@@ -9,7 +9,11 @@ export const getAttributeByPath = (slideAttributes, pathArray: string[]) => {
 
     for (const node of pathArray) {
-        slideAttributes = slideAttributes[node] || slideAttributes[0][node];
+        if (Array.isArray(slideAttributes)) {
+            slideAttributes = slideAttributes[node] || slideAttributes[0][node];
+        } else {
+            slideAttributes = slideAttributes[node];
+        }
```
not sure about this, so if it is array then you access via `slideAttributes[node]` ... but you do the same if it is not array as well .. so is it that this can be an object or an array and we only need to find if it is an array so we can check index 0?

rabi92 replied:
`slideAttributes[node]` is not required if it was an array,  i have removed it. 
I can explain the rest in person regarding the check with array and non array in person if required.

rabi92 committed:
refactor: use constants, reformatting, safety checks

rabi92 committed:
chore: set print width to 120

rabi92 committed:
refactor: unnecessary else and duplication

rabi92 commented:
Not ready for review yet, work in progress. Placeholder images and tables are not getting their positions.

rabi92 committed:
feat: get positions for placeholder pics and tables

refactorization:
- remove code duplications
- make code more generic
- clean up the code
- number of lines reduced by a lot
- code readability
- use of constants
- update version

rabi92 committed:
fix: reject the promise
